crux_core/src/lib.rs

```diff
@@ -191,29 +180,55 @@ where
     ///
     /// The core interface passes across messages serialized as bytes. These can be
     /// deserialized using the types generated using the [typegen] module.
-    pub fn new() -> Self {
-        Self::default()
+    pub fn new<CapFactory>() -> Self
+    where
+        CapFactory: CapabilitiesFactory<A, Ef>,
+    {
+        let (command_sender, command_receiver) = crate::channels::channel();
+        let (event_sender, event_receiver) = crate::channels::channel();
+        let (executor, spawner) = executor::executor_and_spawner();
+        let capability_context = CapabilityContext::new(command_sender, event_sender, spawner);
+
+        Self {
+            model: Default::default(),
+            continuations: Default::default(),
+            executor,
+            app: Default::default(),
+            capabilities: CapFactory::build(capability_context),
+            command_receiver,
+            event_receiver,
+            _marker: PhantomData,
+        }
     }
 
     /// Receive a message from the shell.
     ///
     /// The `msg` is serialized and will be deserialized by the core.
     pub fn message<'de>(&self, msg: &'de [u8]) -> Vec<u8>
     where
-        <A as App<Ef, Caps>>::Event: Deserialize<'de>,
+        <A as App>::Event: Deserialize<'de>,
     {
-        let msg: <A as App<_, _>>::Event =
+        let event: <A as App>::Event =
             bcs::from_bytes(msg).expect("Message deserialization failed.");
 
-        let commands = {
+        let mut model = self.model.write().expect("Model RwLock was poisoned.");
+        self.app.update(event, &mut model, &self.capabilities);
+        drop(model);
+
+        self.executor.run_all();
+
+        while let Some(event) = self.event_receiver.receive() {
```

Comment:
Yeah, I mean _arguably_ you might not need this.  The shell hasn't passed anything in so it is unlikely that any new events have been generated by capabilities.  But I think it would be a mistake to assume that's always true.

Possible that a capability could contain a cache for example and could answer (or partially answer) a request without ever needing to send off an effect.

Reply:
Yeah, I could easily see caching capabilities

Reply:
Yes, ok that all makes sense, and caching capabilities sound like a useful thing